### ISSUE TITLE
Turning all connectivity functions into scikit-learn compatible classes?

### ISSUE THREAD

**Issue by adam2392** [mne-tools/mne-connectivity#21]
#### Describe the problem
Often times, one would like to use connectivity functions for certain regression problems. 

#### Describe your solution
A scikit-learn compatible API would then be able to leverage `GridSearchCV` and `Pipeline` objects in scikit-learn.

#### Additional context
See [MNE-HFO](https://github.com/mne-tools/mne-hfo) for how we enable this for more niche data structure, like decoding events in time. 

I suppose to fit in scikit-learn though, it might be difficult, since I don't think they support "multivariate output" (i.e. ndarrays).

**Comment by larsoner** [comment]
Sounds reasonable to me. In MNE we have `mne.decoding` that wraps some common functional things (such as filtering) to be pipeline compatible. Some new `mne_connectivity.decoding.*` classes that did the same could work well I think.

**Comment by adam2392** [comment]
Perhaps this is naive question:

But would this inherit from a `BaseEstimator`? If so, how would we specify the output correctly, since it's not a `y = (n_samples,)` vector, but a possibly `([n_epochs], n_estimated_nodes, [n_freqs], [n_times])` array?

**Comment by larsoner** [comment]
You need to decide what are "features" and make a `(n_samples, n_features)` shaped output at some point. So for example you probably need to have have `n_epochs` (otherwise what are your multiple samples to actually do anything useful with?) and then you make your output of shape `(n_epochs, n_estimated_nodes * n_freqs * n_times)` via [Vectorizer](https://mne.tools/dev/generated/mne.decoding.Vectorizer.html).

Maybe it would help to first define an example of how you would want to use a connectivity metric (any of them!) in a sklearn pipeline, and then we can hash out an API that would work.

**Comment by agramfort** [comment]
I don't clearly get the use case. The sklearn design is made for pipelining

to me you are here stateless. I would prefer pure functions to be honest.

Then you can use a FunctionTransformer in sklearn to get a sklearn object
from it.

>
